src/server/node_manager.c

```diff
@@ -6850,124 +6895,69 @@ set_nodes(void *pobj, int objtype, char *execvnod_in, char **execvnod_out, char
-			for (j=0; j<psvrmom->msr_jbinxsz; j++)
-				if (psvrmom->msr_jobindx[j] != NULL)
-					log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_JOB, LOG_INFO, pjob->ji_qs.ji_jobid, psvrmom->msr_jobindx[j]->ji_qs.ji_jobid);
-			sprintf(log_buffer, "===================================================================="
-				"==================================================");
-			log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_JOB, LOG_INFO, pjob->ji_qs.ji_jobid, log_buffer);
-		}
+	/*  remove this jobs's jobinfo entry from each vnode   */
+	
+	if (pjob->ji_wattr[(int)JOB_ATR_exec_vnode].at_flags & ATR_VFLAG_SET) {
+		execvnod_in = pjob->ji_wattr[(int)JOB_ATR_exec_vnode].at_val.at_str;
+		if (!execvnod_in)
+			return;
+	} else {
+		log_event(PBSEVENT_DEBUG2, PBS_EVENTCLASS_JOB, LOG_DEBUG,
+		pjob->ji_qs.ji_jobid, "in free_nodes and no exec_vnode");
+		return;
+	}
 
-		for (ivnd = 0; ivnd < psvrmom->msr_numvnds; ++ivnd) {
-			pnode = psvrmom->msr_children[ivnd];
-			still_has_jobs = 0;
-			for (np = pnode->nd_psn; np; np = np->next) {
-
-				for (prev=NULL, jp=np->jobs; jp; jp=next) {
-					next = jp->next;
-					if (jp->job != pjob) {
-						prev = jp;
-						still_has_jobs = 1; /* another job still here */
-						continue;
-					}
+	if (strchr(execvnod_in, (int)':') == NULL) {
+		/* need to take node only list and build a pseudo */
+		/* exec_vnode string with the resources included  */
+		execvnod = build_execvnode(pjob, execvnod_in);
+	} else {
+		execvnod = execvnod_in;
+	}
+	execvncopy = strdup(execvnod);
+	if (execvncopy == NULL)
+		return;
 
-					DBPRT(("Freeing node %s/%ld from job %s\n",
-						pnode->nd_name, np->index,
-						pjob->ji_qs.ji_jobid))
-					if (prev == NULL)
-						np->jobs = next;
-					else
-						prev->next = next;
-					if (jp->has_cpu) {
-						pnode->nd_nsnfree++;	/* up count of free */
-						if (pnode->nd_nsnfree > pnode->nd_nsn) {
-							log_event(PBSEVENT_SYSTEM,
-								PBS_EVENTCLASS_NODE, LOG_ALERT,
-								pnode->nd_name,
-								"CPU count incremented free more than total");
-						}
-					}
-					free(jp);
-					jp = NULL;
-					DBPRT(("%s: upping free count to %ld\n", __func__,
-						pnode->nd_nsnfree))
-				}
-				if (np->jobs == NULL) {
-					np->inuse &= ~(INUSE_JOB|INUSE_JOBEXCL);
-				}
-			}
-			if (still_has_jobs) {
-				/* if the vnode still has jobs, then don't clear JOBEXCL */
-				if (pnode->nd_nsnfree > 0) {
-					/* some cpus free, clear "job-busy" state */
-					set_vnode_state(pnode, ~INUSE_JOB, Nd_State_And);
-				}
-			} else {
-				/* no jobs at all, clear both JOBEXCL and "job-busy" */
-				set_vnode_state(pnode,
-					~(INUSE_JOB|INUSE_JOBEXCL),
-					Nd_State_And);
-
-				/* call function to check and free the node from the prov list
-				 and reset wait_prov flag, if set */
-				if (pjob->ji_qs.ji_substate == JOB_SUBSTATE_PROVISION)
-					free_prov_vnode(pnode);
-			}
+	chunk = parse_plus_spec_r(execvncopy, &last, &hasprn);
```

Comment:
I'm a bit concerned that parse_plus_spec can be slow for long exec_vnodes, can you please also do a performance comparison with multi-node jobs which occupy 100+ nodes and post results before and after your changes?

Reply:
This parse_plus_spec_r() is used extensively in all many places in Server, Sched & MoM. While optimizing this function in general, this code also would get optimized. 

Let me try a multi-node job test also and post the results. Thanks for your review.

Reply:
@agrawalravi90 I tried sampling server with 5000 number of multi-node jobs with each job requesting 100 chunks of 10 cpus. 

The main issue is observed only in single node job submission. It seems fine in multi-node jobs. 

Before fix: 
<img width="1645" alt="Before_fix_multinode" src="https://user-images.githubusercontent.com/17980660/88235077-1c232b80-cc48-11ea-9f1a-1ff5e2891c0e.png">

After fix: 

<img width="1645" alt="After_fix_multinode" src="https://user-images.githubusercontent.com/17980660/88235087-22190c80-cc48-11ea-98ed-39d4c62757d0.png">

Reply:
A few questions:
- How many nodes did you have? how many ncpus on each node?
- I'm a bit confused why req_modifyjob() is taking so much time, especially in the first graph, did you not have enough cpus for ALL jobs to run?
- Can you please highlight free_nodes() in the graphs? You can search for a symbol in perf report just like you do in vim

Reply:
* How many nodes did you have? how many ncpus on each node?
 In my setup, I have 500 vnodes and each node is having 100 cpus each. In total 50k cpus. 
* I'm a bit confused why req_modifyjob() is taking so much time, especially in the first graph, did you not have enough cpus for ALL jobs to run?
 The requested amount cpus are available in the test system
* Can you please highlight free_nodes() in the graphs? You can search for a symbol in perf report just like you do in vim
 I am not able to locate the free_nodes() while running in multi-node tests. I have tried various test setups, Its happening mostly in single node tests. Even searched in vim too.

Reply:
Please ignore the previous results:

I have tried the setup suggested by @agrawalravi90  (Thanks) , In this setup, I could get the free_nodes symbol perfectly. 

Setup details: 

1) 50000 vnodes, each with 10 cpus. 
2) Submitted 5k jobs, each requesting 10 chunks of 10 ncpus,

Scanned the server in perf till the job completion with one scheduling cycle. 

Before fix:
<img width="1645" alt="master_multinode_50knodes" src="https://user-images.githubusercontent.com/17980660/88244627-5a2d4900-cc62-11ea-8cec-0dd1ff39a357.png">

After fix: 
<img width="1645" alt="master_after_fix_50knodes" src="https://user-images.githubusercontent.com/17980660/88244635-60232a00-cc62-11ea-9e5f-fd027769bd32.png">